mongoimport/common.go

```diff
@@ -399,28 +570,204 @@ func validateFields(fields []string) error {
+// with addFieldToArray(). It closely mimics the behaviour of setNestedDocumentValue(). See validateFields()
+// for more information on the validity checks that are made when constructing a tree of fields.
+func addFieldToTree(fieldParts []string, fullField string, fieldPrefix string, tree map[string]interface{}, useArrayIndexFields bool) (map[string]interface{}, error) {
+	head, tail := fieldParts[0], fieldParts[1:]
+	if fieldPrefix == "" {
+		fieldPrefix = head
+	} else {
+		fieldPrefix += "." + head
+	}
+
+	value, exists := tree[head]
+
+	if exists && len(fieldParts) == 1 {
+		if value == true {
+			// fields are the same - case (2)
+			return nil, identicalError(fullField)
+		}
+		// case (3)
+		return nil, incompatibleError(fullField, fieldPrefix, value)
+	}
+
+	if len(fieldParts) == 1 {
+		tree[head] = true
+		return tree, nil
+	}
+
+	// Now determine the type of the next part of the field
+
+	if _, ok := isNatNum(tail[0]); useArrayIndexFields && ok {
+		var subArray []interface{}
+		if exists {
+			subArray, ok = value.([]interface{})
+			if !ok {
+				// case (5) or (4)
+				return nil, incompatibleError(fullField, fieldPrefix, value)
 			}
-			// NOTE: this means we will not support imports that have fields like
-			// a, a - since this is invalid in MongoDB
-			if field == latterField {
-				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, latterField)
+		} else {
+			subArray = make([]interface{}, 0)
+		}
+		subArray, err := addFieldToArray(tail, fullField, fieldPrefix, subArray)
+		if err != nil {
+			return nil, err
+		}
+		tree[head] = subArray
+	} else {
+		var subTree map[string]interface{}
+		if exists {
+			subTree, ok = value.(map[string]interface{})
+			if !ok {
+				// case (5) or (4)
+				return nil, incompatibleError(fullField, fieldPrefix, value)
 			}
+		} else {
+			subTree = make(map[string]interface{})
 		}
+		subTree, err := addFieldToTree(tail, fullField, fieldPrefix, subTree, useArrayIndexFields)
+		if err != nil {
+			return nil, err
+		}
+		tree[head] = subTree
 	}
-	return nil
+	return tree, nil
+
+}
+
+// addFieldToArray is used with addFieldToTree() to build a valid tree of fields.
+func addFieldToArray(fieldParts []string, fullField string, fieldPrefix string, array []interface{}) ([]interface{}, error) {
+	head, tail := fieldParts[0], fieldParts[1:]
+	fieldPrefix += "." + head
+
+	// The first part of the field should be an index of an array
+	headIndex, ok := isNatNum(head)
+	if !ok {
+		// We shouldn't ever get here
+		panic(fmt.Sprintf("addFieldToArray expected an integer field, but instead received %s", fieldParts[0]))
+	}
+
+	if len(fieldParts) == 1 {
+		// We're at the terminus of a field so we have to check if we can append to the array
+		if headIndex == len(array) {
+			array = append(array, true)
+			return array, nil
+		}
+		return nil, indexError(fullField)
+	}
+
+	if headIndex > len(array) {
```

Comment:
Suggest adding a brief comment  above explaining this case in English. (e.g. "headIndex must be an existing element or be adding an element at the end"?)